Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,815 @@
+        // as long as we don't show the fade happening
+        if (cmd->length == 7)
+        {
+          data = {buf[0], buf[1], buf[2], buf[3], buf[4], buf[5], buf[6]};
+          expected_count = 15;
+          result = {buf[0]};
+          m_queries.push(result);
+          g_skyportal.SetLEDs(buf[1], buf[2], buf[3], buf[4]);
+        }
+        break;
+      }
+      case 'L':
+      {
+        // Light
+        // This command is used while playing audio through the portal
+
+        // The 2nd bytes is the position
+        // 0x00: right
+        // 0x01: trap led
+        // 0x02: left
+
+        // The 3rd, 4th and 5th bytes are red, green and blue
+        // the trap led is white-only
+        // increasing or decreasing the values results in a brighter or dimmer light
+        if (cmd->length == 5)
+        {
+          data = {buf[0], buf[1], buf[2], buf[3], buf[4]};
+          expected_count = 13;
+
+          u8 side = buf[1];
+          if (side == 0x02)
+          {
+            side = 0x04;
+          }
+          g_skyportal.SetLEDs(side, buf[2], buf[3], buf[4]);
+        }
+        break;
+      }
+      case 'M':
+      {
+        // Audio Firmware version
+        // Respond with version obtained from Trap Team wired portal
+        if (cmd->length == 2)
+        {
+          data = {buf[0], buf[1]};
+          expected_count = 10;
+          result = {buf[0], buf[1], 0x00, 0x19};
+          m_queries.push(result);
+        }
+        break;
+      }
+        // Query
+        // Command	{ 'Q', 10, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
+        // 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00 }
+        // Response	{ 'Q', 10, 00, 00, 00, 00,
+        // 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
+        // 00, 00, 00, 00 }
+        // In the command the 2nd byte indicates which Skylander to query data
+        // from. Index starts at 0x10 for the 1st Skylander (as reported in the Status command.) The
+        // 16th Skylander indexed would be 0x20. The 3rd byte indicate which block to read from.
+
+        // A response with the 2nd byte of 0x01 indicates an error in the read. Otherwise, the
+        // response indicates the Skylander's index in the 2nd byte, the block read in the 3rd byte,
+        // data (16 bytes) is contained in bytes 4-19.
+
+        // A Skylander has 64 blocks of data indexed from 0x00 to 0x3f. SwapForce characters have 2
+        // character indexes, these may not be sequential.
+      case 'Q':
+      {
+        // Queries a block
+        if (cmd->length == 3)
+        {
+          const u8 sky_num = buf[1] & 0xF;
+          const u8 block = buf[2];
+          g_skyportal.QueryBlock(sky_num, block, result.data());
+          m_queries.push(result);
+          data = {buf[0], buf[1], buf[2]};
+          expected_count = 11;
+          break;
```

Comment:
This break should probably be outside the if statement. Otherwise, an invalid length will cause control flow to fall through the next case